samza-core/src/main/java/org/apache/samza/zk/ScheduleAfterDebounceTime.java

```diff
@@ -51,10 +50,23 @@
 
   public static final int DEBOUNCE_TIME_MS = 2000;
 
+  private final ScheduledTaskFailureCallback scheduledTaskFailureCallback;
+
   private final ScheduledExecutorService scheduledExecutorService = Executors.newSingleThreadScheduledExecutor(
       new ThreadFactoryBuilder().setNameFormat("debounce-thread-%d").setDaemon(true).build());
   private final Map<String, ScheduledFuture> futureHandles = new HashMap<>();
 
+  // Ideally, this should be only used for testing. But ZkBarrierForVersionUpgrades uses it. This needs to be fixed.
+  // TODO: Timer shouldn't be passed around the components. It should be associated with the JC or the caller of
+  // coordinationUtils.
+  public ScheduleAfterDebounceTime() {
```

Comment:
I think the debounceTimer should be associated with the groupId of the coordinationUtils, so the same thread is used for all the action related to the same processor. Also I don't where this constructor is invoked, so may be it can be removed. 
On the other hand, it looks like errorHandling listener should be passed from the point of invocation, so it could be different for different components.